include/gridtools/fn/extents.hpp

```diff
@@ -0,0 +1,113 @@
+/*
+ * GridTools
+ *
+ * Copyright (c) 2014-2021, ETH Zurich
+ * All rights reserved.
+ *
+ * Please, refer to the LICENSE file in the root directory.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#pragma once
+
+#include <algorithm>
+#include <cstddef>
+#include <type_traits>
+
+#include "../common/int_vector.hpp"
+#include "../common/integral_constant.hpp"
+#include "../common/utility.hpp"
+#include "../meta.hpp"
+#include "gridtools/meta/type_traits.hpp"
+
+namespace gridtools {
+    namespace fn {
+        template <class Dim, ptrdiff_t L, ptrdiff_t U>
+        struct extent {
+            static_assert(L <= U);
+            using dim_t = Dim;
+            using lower_t = integral_constant<ptrdiff_t, L>;
+            using upper_t = integral_constant<ptrdiff_t, U>;
+            using size_t = integral_constant<std::size_t, U - L>;
+            using list_t = meta::list<Dim, lower_t, upper_t>;
+        };
+
+        template <class>
+        struct is_extent : std::false_type {};
+
+        template <class Dim, ptrdiff_t L, ptrdiff_t U>
+        struct is_extent<extent<Dim, L, U>> : bool_constant<L <= U> {};
+
+#ifdef __cpp_concepts
+        template <class T>
+        concept extent_c = is_extent<T>::value;
+#endif
+
+        template <class... Ts>
+        struct extents {
+            static_assert(meta::all_of<is_extent, meta::list<Ts...>>::value);
+            static_assert(meta::is_set<meta::list<typename Ts::dim_t...>>::value);
+            using keys_t = hymap::keys<typename Ts::dim_t...>;
+            static_assert(is_int_vector<typename keys_t::template values<typename Ts::lower_t...>>::value);
+            static_assert(is_int_vector<typename keys_t::template values<typename Ts::upper_t...>>::value);
+            static_assert(is_int_vector<typename keys_t::template values<typename Ts::size_t...>>::value);
+
+            static GT_CONSTEVAL auto offsets() {
+                return int_vector::prune_zeros(typename keys_t::template values<typename Ts::lower_t...>());
+            }
+            static GT_CONSTEVAL auto sizes() {
+                return int_vector::prune_zeros(typename keys_t::template values<typename Ts::size_t...>());
+            }
+        };
+
+        template <class>
+        struct is_extents : std::false_type {};
+
+        template <class... Ts>
+        struct is_extents<extents<Ts...>> : bool_constant<meta::all_of<is_extent, meta::list<Ts...>>::value &&
+                                                          meta::is_set<meta::list<typename Ts::dim_t...>>::value> {};
+
+#ifdef __cpp_concepts
+        template <class T>
+        concept extents_c = meta::is_instantiation_of<extents, T>::value;
```

Comment:
In that form `extents_c` concept is not consistent with `is_extents`. I would suggest to remove it completely and add TODO.   Otherwise we need to test that iff `is_extents` is true `extents_c` is also true.

Reply:
I replaced by
```
        template <class T>
        concept extents_c = is_extents<T>::value;
```
I believe in this trivial form it doesn't harm to leave it in.